pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,121 @@
+"""
+# pylint: disable=protected-access, too-many-arguments
+
+
+import pennylane as qml
+from pennylane.operation import StatePrepBase
+from pennylane.tape import QuantumScript
+from pennylane.transforms.core import transform
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def decompose(tape, gate_set=None, max_expansion=None):
+    """Decomposes operations found in a quantum circuit into a desired gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (Callable or set, optional): A set of decomposition gates or a condition they satisfy. Defaults to None.
+        max_expansion (int, optional): The maximum depth of the expansion. Defaults to None.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    Raises:
+        Exception: Type defaults to ``qml.DeviceError``.
+            Raised if an operator is not accepted and does not define a decomposition, or if
+            the decomposition enters an infinite loop and raises a ``RecursionError``.
+
+    .. seealso:: :func:`~.pennylane.devices.preprocess.decompose` for a transform that performs the same job but is designed particularly for specific device architectures.
+
+    **Examples:**
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
+    >>> batch, fn = decompose(tape, gate_set = {"CNOT", "RX", "RZ"})
+    >>> batch[0].circuit
+    [CNOT(wires=[0, 1]),
+    RX(1.2, wires=[0]),
+    CNOT(wires=[0, 1]),
+    expval(Z(0))]
+
+    >>> @partial(decompose, gate_set = lambda obj: len(obj.wires) <= 2)
+    >>> @qml.qnode(device)
+    >>> def circuit():
+    >>>     qml.Toffoli(wires = range(2))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ───────────╭●───────────╭●────╭●──T──╭●─┤
+    1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤
+    2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤
+
+    """
+    if gate_set is None:
+        gate_set = set(qml.ops.__all__)
+
+    if isinstance(gate_set, (list, tuple)):
+        gate_set = set(gate_set)
+
+    def decomposer(op):
+        return op.decomposition()
+
+    def stopping_condition(op):
+        if not isinstance(op, qml.operation.Operator):
+            return True
+        if not op.has_decomposition:
+            return True
+        if isinstance(gate_set, set):
+            return op.name in gate_set
+        return gate_set(op)
+
+    if tape.operations and isinstance(tape[0], StatePrepBase):
+        prep_op = [tape[0]]
+    else:
+        prep_op = []
```

Comment:
I don't think this makes much sense without the `skip_initial_state_prep` argument, and I don't think that's needed here - users can just include StatePrep in their supported gates if they are happy with it not being decomposed.